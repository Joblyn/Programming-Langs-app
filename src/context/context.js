import { createContext } from "react";

export const RouteContext = createContext(null);

export const NameContext = createContext(null);