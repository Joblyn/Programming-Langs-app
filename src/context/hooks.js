export const useAppContext = (value = null) => {
  let route = value;
  return route;
};