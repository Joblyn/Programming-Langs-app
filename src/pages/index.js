export { default as Home } from './home';
export { default as Javascript } from './javascript';
export { default as Python } from './python';
export { default as Java } from "./java";