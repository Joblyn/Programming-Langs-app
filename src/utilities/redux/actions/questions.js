import { createAction } from "@reduxjs/toolkit";
import * as ActionTypes from "../../../constants/actionTypes";
import { getData } from "../../services";

const onGetQuestions = (payload) => {
  let getQuestionsAction =
    payload.status === 200
      ? createAction(ActionTypes.GET_QUESTIONS)
      : createAction(ActionTypes.GET_QUESTIONS_ERROR);

  return getQuestionsAction(payload);
};

export const getQuestions = (endpoint) => {
  return getData(endpoint, onGetQuestions);
};
