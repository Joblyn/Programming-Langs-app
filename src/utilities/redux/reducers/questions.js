import { questions } from "../initialState";
import * as ActionTypes from "../../../constants/actionTypes";

export const questionsReducer = (state = questions, action) => {
  switch (action.type) {
    case ActionTypes.GET_QUESTIONS:
      return {
        isSuccessful: true,
        isError: false,
        data: action.payload.data,
        error: null,
      };
    case ActionTypes.GET_QUESTIONS_ERROR:
      return {
        isSuccessful: false,
        isError: true,
        data: [],
        error: action.payload.error,
      };
    default:
      return state;
  }
};
