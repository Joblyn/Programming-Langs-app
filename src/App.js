import React, { useState, useEffect } from "react";
import "./App.css";
import { Routes, Route } from "react-router-dom";
import * as ROUTES from "./constants/routes";
import { Home, PageNotFound, Assessment, Name } from "./pages";
import Layout from "./containers/layout/layout";
import { AppContext } from "./context/context";
import ReactGA from "react-ga";

// initialize google analytics
let TRACKING_ID;
if (window.location.hostname === "woshipfrontend.interviewblindspots.com") {
  TRACKING_ID = "G-QW0ZP2JZHF";
} else {
  if (window.location.hostname === "damp-shore-08296.herokuapp.com") {
    TRACKING_ID = "G-L3SYMZVBRR";
  } else {
    TRACKING_ID = "G-8WCMDG75N4";
  }
}
ReactGA.initialize(TRACKING_ID);

function App() {
  const [route, setRoute] = useState();
  const [name, setName] = useState();

  useEffect(() => {
    ReactGA.pageview(window.location.pathname + window.location.search);
  }, []);

  return (
    <AppContext.Provider value={{ route, setRoute, setName, name }}>
      <Routes>
        <Route index element={<Home />} />
        <Route path={ROUTES.HOME} element={<Home />} />
        <Route path={ROUTES.ASSESSMENT} exact element={<Layout />}>
          <Route path={ROUTES.NAME} exact element={<Name />} />
          <Route path={ROUTES.LANGUAGE} exact element={<Assessment />} />
        </Route>
        <Route path={ROUTES.NOMATCH} element={<PageNotFound />} />
      </Routes>
    </AppContext.Provider>
  );
}

export default App;
