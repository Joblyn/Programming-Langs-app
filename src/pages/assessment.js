import React, { useState, useEffect, createElement } from "react";
import NameIcon from "../assets/images/name_icon.svg";
import ProgressLine from "../components/progress/progressline";
import { LargeButton } from "../components/button/button";
import Question from "../components/question/question";
import { useMediaQuery } from "@mui/material";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { getQuestions } from "../utilities/redux/actions/questions";
import Loader from "../components/loader/loader";

export default function Assessment() {
  const { language } = useParams();
  const [count, updateCount] = useState(1);
  const [startEnd, setStartEnd] = useState({ start: 0, end: 5 });
  const [data, setData] = useState([]);

  const [last, setLast] = useState(false);
  const remainder = data.length % 5;
  const quotient = parseInt(data.length / 5);
  const matches = useMediaQuery("(min-width:960px)");
  const dispatch = useDispatch();

  const questions = useSelector((state) => state.questions);
  const [questionsState, setQuestionsState] = useState([]);

  const responseStore = useSelector((state) => state.responses);
  const [responses, setResponses] = useState({});

  const [error, setError] = useState({});

  // dispatch action to fetch questions and store at redux store
  useEffect(() => {
    dispatch(getQuestions());
  }, [language, dispatch]);

  // store all language questions from redux store
  useEffect(() => {
    if (questions.isSuccessful) {
      setData(questions.data);
    }
    if (questions.isError) {
      console.error()
      setError({ type: "fetch error", message: error });
    }
  }, [questions, error]);

  // set current questions and responses sets
  useEffect(() => {
    if (startEnd.end === undefined) {
      setQuestionsState(data.slice(startEnd.start));
    } else {
      setQuestionsState(data.slice(startEnd.start, startEnd.end));
    }
  }, [data, startEnd]);

  // check if current questions set is the last and update state
  useEffect(() => {
    if (remainder === 0) {
      if (count === quotient) {
        setLast(true);
      } else {
        setLast(false);
      }
      return;
    }
  }, [count, remainder, quotient]);

  // update responses state
  useEffect(() => {
    setResponses(function () {
      let obj = {};
      questionsState.forEach((question) => {
        let prop = `question_${question.id}`;
        obj = { ...obj, [prop]: responseStore[prop] || "" };
      });
      return obj;
    });
  }, [questionsState, responseStore]);

  const nextQuestionsSet = () => {
    updateCount((count) => count + 1);
    if (remainder) {
      if (count === quotient) {
        setStartEnd(({ end }) => ({ start: end }));
        setLast(true);
        setError({});
      }
    } else {
      setStartEnd(({ end }) => ({ start: end, end: end + 5 }));
      setError({});
    }
  };

  const previousQuestionSet = () => {
    updateCount((count) => count - 1);
    setStartEnd(({ start, end }) => ({ start: start - 5, end: end - 5 }));
    setError({});
  };

  const createCommponent = () => {
    const getProps = (questions, id) => ({
      className: id === questions.length - 1 ? "pb-[173px] md:pb-[100px] question_container" : "question_container",
      key: "question_" + id,
    });
    return questionsState.map((question, id) => {
      let propName = "question_" + question.id;
      return createElement(
        "div",
        getProps(questionsState, id),
        <Question
          propName={propName}
          id={id}
          defaultValue={responseStore[propName]}
          question={question}
          setResponses={setResponses}
        />
      );
    });
  };

  let incompleteResponses = Object.values(responses).includes("");

  return (
    <div className="w-full relative" id="top">
      <div className="flex items-center my-2 w-full">
        <span className="align-middle mr-3">
          <img src={NameIcon} alt="_logo_" className="w-10 h-10" />
        </span>
        <span className="uppercase text-green text-lg font-bold ">
          {`${language} assessment`}
        </span>
      </div>
      {questionsState.length ? (
        <>
          <ProgressLine count={count} totalcount={quotient} />
          {createCommponent()}
          <div className={`${matches ? "px-6 pb-6" : "assessment_footer"}`}>
            {error.type === "incomplete responses" && (
              <p className="text-[red] text-base px-6">{error.message}</p>
            )}
            <div className="flex gap-3">
              {count !== 1 && (
                <LargeButton
                  type="back"
                  innerText={"Back"}
                  action={previousQuestionSet}
                />
              )}
              <LargeButton
                type="next"
                innerText={last ? "Finish" : "Next"}
                action={
                  incompleteResponses
                    ? () =>
                        setError({
                          type: "incomplete responses",
                          message: "Please fill in all fields",
                        })
                    : last
                    ? null
                    : nextQuestionsSet
                }
              />
            </div>
          </div>
        </>
      ) : (
        <Loader />
      )}
    </div>
  );
}
