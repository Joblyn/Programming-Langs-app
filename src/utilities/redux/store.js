import { configureStore } from "@reduxjs/toolkit";
import errorReducer from "./reducers/error";
import { questionsReducer } from "./reducers/questions";
import { responsesReducer } from "./reducers/responses";

export const store = configureStore({
  reducer: {
    questions: questionsReducer,
    responses: responsesReducer,
    connectionError: errorReducer
  }
})