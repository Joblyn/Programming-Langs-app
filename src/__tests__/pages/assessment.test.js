import React from "react";
import { render, cleanup, fiireEvent, fireEvent } from "@testing-library/react";
import { createMemoryHistory } from "history";
// import { Router } from "react-router-dom";
import { Provider } from "react-redux";
import "@testing-library/jest-dom";

import { Assessment } from "../../pages";
import { AppContext } from "../../context/context";
import App from "../../App";
import { getQuestions } from "../../utilities/redux/actions/questions";
import { store } from "../../utilities/redux/store";
import { act } from "react-test-renderer";

// test the app at assessment page
describe("<Assessment />", () => {
  afterEach(cleanup);

  // mock react-router-dom to return params
  // jest.mock("react-router-dom", () => ({
  //   ...jest.requireActual("react-router-dom"),
  //   useParams: () => ({
  //     language: "java",
  //   }),
  // }));

  test("renders loader when questions are not available", () => {
    // mock redux store
    const mockDispatch = jest.fn(() => null);
    jest.mock("react-redux", () => ({
      useDispatch: () => mockDispatch,
    }));

    // change pathname to assessment/:language
    const history = createMemoryHistory();
    history.push("/assessment/java");
    const { container, getByTestId, getByText } = render(
      // <Router location={history.location} navigator={history}>
      <Provider store={store}>
        <Assessment />
      </Provider>
      // </Router>
    );

    // const getQuestionsResult = getQuestions();
    // expect(getQuestionsResult).toBe(null);

    // expect a header with text stating the language assessment
    // expect(getByText("Java assessment", { exact: false }));

    // component renders loader when questions are not available
    expect(getByTestId("loader-container")).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders questions when questions are ready", async () => {
    // mock getQuestions function to return null value
    // const getQuestions = jest.fn();
    // jest.mock("../../utilities/redux/actions/questions", () => {
    //   const originalModule = jest.requireActual(
    //     "../../utilities/redux/actions/questions"
    //   );
    //   return {
    //     _esModule: true,
    //     ...originalModule,
    //     getQuestions: () => getQuestions,
    //   };
    // });

    const questionsData = Array.from(new Array(10), (val, id) => ({
      id: id + 1,
      text: "Do you understand Javascript objects?",
      choices: "Yes|No|Maybe",
    }));
    const mockDispatch = jest.fn(() => null);
    const mockUseSelector = jest.fn().mockReturnValue({
        isSuccessful: true,
        isError: false,
        error: null,
        data: questionsData,
      }
    )
    jest.mock("react-redux", () => {
      return {
        ...jest.requireActual("react-redux"),
        useSelector: () => mockUseSelector,
        useDispatch: () => mockDispatch,
      };
    });

    const { container, getByText, getByTestId } = render(
      <Provider store={store}>
        <Assessment />
      </Provider>
    );

    const mockQuestions = mockUseSelector(state => state.questions);
    expect(mockQuestions).toBe();
    // expect to see 5 questionswf on the page
    expect(container.getElementsByClassName("question_container").length).toBe(5);
    // expect questions to be numbered
    expect(container.getElementsByClassName("question_number")).toBe(5);
    // expect to see options
    expect(
      container
        .getElementsByClassName("question_container")[0]
        .contains(".opt-cont")
    ).toBeTruthy();
    // expect to progress count to be 1/2
    expect(getByText("1/2")).toBeTruthy();

    // user cannot move to next set without all options being filled, show an error message
    await fireEvent.click(getByText(/next/));
    expect(getByText(/Please fill in all fields/)).toBeTruthy();
    expect(getByText("1/2")).toBeTruthy();

    // await act(async () => {
    //   await fireEvent.change()
    // });
  });
});
// render with store
// component calls getQuestions() once rendered
// questions are updated
// component renders questions with options
// check that count is 1
// user cannot click next without options being filled
// click an option for one question
// next button should be disabled
// click an option for every question
// next button should enabled
// user clicks next button
// check new set of questions
// check that button cannot be clicked
// check that count is 2
// click back button
// check that count is 1
// check that next button is not disabled
// click next button
// check that count is 2
