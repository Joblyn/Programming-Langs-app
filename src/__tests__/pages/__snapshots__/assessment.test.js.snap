// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Assessment /> renders loader when questions are not available 1`] = `
<div
  class="w-full relative"
  id="top"
>
  <div
    class="flex items-center my-2 w-full"
  >
    <span
      class="align-middle mr-3"
    >
      <img
        alt="_logo_"
        class="w-10 h-10"
        src="name_icon.svg"
      />
    </span>
    <span
      class="uppercase text-green text-lg font-bold "
    >
      undefined assessment
    </span>
  </div>
  <div
    class="h-[80vh] w-full relative flex items-center justify-center"
    data-testid="loader-container"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSuccess css-15u8j6q-MuiCircularProgress-root"
      role="progressbar"
      style="width: 25px; height: 25px;"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</div>
`;
